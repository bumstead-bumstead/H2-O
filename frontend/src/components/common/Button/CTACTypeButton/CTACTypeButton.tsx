import { ButtonHTMLAttributes } from 'react';
import styled from '@emotion/styled';

type ButtonSize = 'small' | 'large';
type ButtonColor = 'gray' | 'primary';

interface Props extends ButtonHTMLAttributes<HTMLButtonElement> {
  size?: ButtonSize;
  color?: ButtonColor;
}

function CTACTypeButton({ children, size = 'small', color = 'gray', ...restProps }: Props) {
  return (
    <StyledButton size={size} color={color} {...restProps}>
      {children}
    </StyledButton>
  );
}

export default CTACTypeButton;

const StyledButton = styled.button<Pick<Props, 'size' | 'color'>>`
  ${({ theme }) => theme.flex.flexCenterRow}
  ${({ theme }) => theme.typography.HeadKRMedium16}
  background-color: ${({ theme, color }) => (color === 'gray' ? theme.colors.gray300 : theme.colors.primary700)};
  color: ${({ theme }) => theme.colors.white};

  width: ${({ size }) => (size === 'small' ? `168px` : `336px`)};
  height: 52px;

  &:active {
    background-color: ${({ theme, color }) => (color === 'gray' ? theme.colors.gray400 : theme.colors.primary800)};
  }
`;
