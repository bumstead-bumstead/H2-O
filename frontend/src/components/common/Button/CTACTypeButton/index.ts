export { default as CTACTypeButton } from './CTACTypeButton';
