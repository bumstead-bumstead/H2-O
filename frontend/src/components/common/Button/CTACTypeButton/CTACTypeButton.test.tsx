import CTACTypeButton from './CTACTypeButton';
import { render, screen, userEvent, waitFor } from '@/tests/test-util';

describe('CTAButton 컴포넌트 테스트', () => {
  it('버튼이 정상적으로 렌더링된다.', () => {
    render(<CTACTypeButton>버튼</CTACTypeButton>);

    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  it('버튼의 onClick 핸들러가 클릭시 정상적으로 호출된다.', async () => {
    const mockOnClick = vi.fn();
    render(<CTACTypeButton onClick={mockOnClick}>버튼</CTACTypeButton>);

    userEvent.click(screen.getByRole('button'));
    await waitFor(() => expect(mockOnClick).toBeCalledTimes(1));
  });

  it('버튼에 disabled 속성을 전달하면 disabled 된다.', async () => {
    const mockOnClick = vi.fn();
    render(
      <CTACTypeButton disabled onClick={mockOnClick}>
        버튼
      </CTACTypeButton>,
    );

    const button = screen.getByRole('button');
    expect(button).toBeDisabled();

    userEvent.click(button);
    await waitFor(() => expect(mockOnClick).not.toBeCalled());
  });

  describe('CTACTypeButton 스냅샷 테스트', () => {
    it('size 속성이 small일 때 스냅샷이 일치한다.', () => {
      render(<CTACTypeButton size='small'>버튼</CTACTypeButton>);

      expect(screen.getByRole('button')).toMatchSnapshot();
    });

    it('size 속성이 large일 때 스냅샷이 일치한다.', () => {
      render(<CTACTypeButton size='large'>버튼</CTACTypeButton>);

      expect(screen.getByRole('button')).toMatchSnapshot();
    });

    it('color 속성이 gray일 때 스냅샷이 일치한다.', () => {
      render(<CTACTypeButton color='gray'>버튼</CTACTypeButton>);

      expect(screen.getByRole('button')).toMatchSnapshot();
    });

    it('color 속성이 primary 일 때 스냅샷이 일치한다.', () => {
      render(<CTACTypeButton color='primary'>버튼</CTACTypeButton>);

      expect(screen.getByRole('button')).toMatchSnapshot();
    });
  });
});
