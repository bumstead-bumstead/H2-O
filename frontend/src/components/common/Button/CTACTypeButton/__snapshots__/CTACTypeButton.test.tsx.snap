// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CTAButton 컴포넌트 테스트 > CTACTypeButton 스냅샷 테스트 > color 속성이 gray일 때 스냅샷이 일치한다. 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  letter-spacing: -0.03em;
  background-color: #C8CACC;
  color: #FFFFFF;
  width: 168px;
  height: 52px;
}

.emotion-0:active {
  background-color: #B6B8BB;
}

<button
  class="emotion-0"
  color="gray"
>
  버튼
</button>
`;

exports[`CTAButton 컴포넌트 테스트 > CTACTypeButton 스냅샷 테스트 > color 속성이 primary 일 때 스냅샷이 일치한다. 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  letter-spacing: -0.03em;
  background-color: #00224A;
  color: #FFFFFF;
  width: 168px;
  height: 52px;
}

.emotion-0:active {
  background-color: #001731;
}

<button
  class="emotion-0"
  color="primary"
>
  버튼
</button>
`;

exports[`CTAButton 컴포넌트 테스트 > CTACTypeButton 스냅샷 테스트 > size 속성이 large일 때 스냅샷이 일치한다. 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  letter-spacing: -0.03em;
  background-color: #C8CACC;
  color: #FFFFFF;
  width: 336px;
  height: 52px;
}

.emotion-0:active {
  background-color: #B6B8BB;
}

<button
  class="emotion-0"
  color="gray"
>
  버튼
</button>
`;

exports[`CTAButton 컴포넌트 테스트 > CTACTypeButton 스냅샷 테스트 > size 속성이 small일 때 스냅샷이 일치한다. 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  letter-spacing: -0.03em;
  background-color: #C8CACC;
  color: #FFFFFF;
  width: 168px;
  height: 52px;
}

.emotion-0:active {
  background-color: #B6B8BB;
}

<button
  class="emotion-0"
  color="gray"
>
  버튼
</button>
`;
